pkg/controller/healthzcheck/controller.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package healthzcheck
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/record"
+	unversioned_core "k8s.io/kubernetes/pkg/client/typed/generated/core/unversioned"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+type APIServerHealthChecker interface {
+	Check(masterAddress string, masterPorts []api.EndpointPort) (bool, error)
+}
+
+type httpsHealthzChecker struct {
+	healthzPath string
+	client      *http.Client
+}
+
+func (c *httpsHealthzChecker) Check(masterAddress string, masterPorts []api.EndpointPort) (bool, error) {
+	if len(masterPorts) == 0 {
+		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no ports provided", masterAddress)
+	}
+
+	masterPort := -1
+	for _, port := range masterPorts {
+		if port.Name == "https" && port.Protocol == api.ProtocolTCP {
+			masterPort = port.Port
+			break
+		}
+	}
+
+	if masterPort == -1 {
+		return false, fmt.Errorf("unable to determine port to use to check API server %s healthz: no TCP port was called 'https'", masterAddress)
+	}
+
+	url := fmt.Sprintf("https://%s:%v/%s", masterAddress, masterPort, c.healthzPath)
```

Comment:
> This needs to be using the Probe classes

As in `pkg/probe/http`?  IMO, it doesn't look quite like what we want:

- The health checker interface should expect to receive a master IP, not be required to give a full URL (that way, different checkers could implement different checks in the future)

- We want to use the client information supplied to kube (using the right client certs, trusting the right CAs, etc) -- the HTTPProber constructor just constructs a plain HTTP client `InsecureSkipVerify` and `DisableKeepAlives` set to true.

Using the `DoHTTPProbe` function instead of manually running `NewRequest` and `client.Do` seems to make sense, though.